src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,506 @@
+    //   "functionName" -> fault config
+    //   "CUPT callback id" -> fault config
+    //   "*" -> fault config
+    const std::string driverFaultsKey = "cudaDriverFaults";
+    const std::string runtimeFaultsKey = "cudaRuntimeFaults";
+
+
+    PTHREAD_CALL(pthread_rwlock_wrlock(&globalControl.configLock));
+    try {
+        boost::property_tree::read_json(jsonStream, globalControl.configRoot);
+        const int logLevel = globalControl.configRoot
+            .get_optional<int>(logLevelKey)
+            .value_or(0);
+
+        globalControl.dynamic = globalControl.configRoot
+            .get_optional<bool>(dynamicConfigKey)
+            .value_or(false);
+
+        const spdlog::level::level_enum logLevelEnum = static_cast<spdlog::level::level_enum>(logLevel);
+        spdlog::info("changed log level to {}", logLevelEnum);
+        spdlog::set_level(logLevelEnum);
+        traceConfig(globalControl.configRoot);
+
+        globalControl.driverFaultConfigs = globalControl.configRoot
+            .get_child_optional(driverFaultsKey);
+        globalControl.runtimeFaultConfigs = globalControl.configRoot
+            .get_child_optional(runtimeFaultsKey);
+    } catch (boost::property_tree::json_parser::json_parser_error& error) {
+        spdlog::error("error parsing fault injector config, still editing? {}", error.what());
+    }
+    PTHREAD_CALL(pthread_rwlock_unlock(&globalControl.configLock));
+    jsonStream.close();
+    spdlog::debug("readFaultInjectorConfig from {} DONE", configFilePath);
+}
+
+static void
+traceConfig(boost::property_tree::ptree const& pTree) {
+    boost::property_tree::ptree::const_iterator end = pTree.end();
+    for (boost::property_tree::ptree::const_iterator it = pTree.begin(); it != end; ++it) {
+        spdlog::trace("congig key={} value={}",  it->first, it->second.get_value<std::string>());
+        traceConfig(it->second);
+    }
+}
+
+static int
+eventCheck(int fd) {
+    fd_set rfds;
+    FD_ZERO(&rfds);
+    FD_SET(fd, &rfds);
+    struct timeval tv;
+    tv.tv_sec = 5;
+    tv.tv_usec = 0;
+    return select(FD_SETSIZE, &rfds, NULL, NULL, &tv);
+}
+
+
+static void *
+dynamicReconfig(void *args) {
+    spdlog::debug("config watcher thread: inotify_init()");
+    const int inotifyFd = inotify_init();
+    if (inotifyFd < 0) {
+        spdlog::error("inotify_init() failed");
+        return NULL;
+    }
+    spdlog::debug("config watcher thread: inotify_add_watch {}", configFilePath);
+    const int watchFd = inotify_add_watch(inotifyFd, configFilePath, IN_MODIFY);
+    if (watchFd < 0) {
+        spdlog::error("config watcher thread: inotify_add_watch {} failed", configFilePath);
+        return NULL;
+    }
+
+    constexpr auto MAX_EVENTS = 1024;
+    const auto configFilePathStr = std::string(configFilePath);
+    constexpr auto EVENT_SIZE = sizeof(struct inotify_event);
+    const auto BUF_LEN = MAX_EVENTS * (EVENT_SIZE + configFilePathStr.length());
+    char eventBuffer[BUF_LEN];
```

Comment:
Please group `constexpr` and other variables in different groups, don't mix them.